### PR TITLE
search frontend: warn on structural search without lang: filter

### PR DESCRIPTION
**Pull request by rvantonder**
This PR adds a warning diagnostic when structural search is active but `lang:` is not set.

When structural search is active users should add a `lang:` filter so that we can pick the right parser. If `lang:` isn't set, we may fall back to a generic parser, which can "miss" results that the user would normally expect. 

More context: https://sourcegraph.slack.com/archives/C07KZF47K/p1667385583731439

## Test plan
Updated unit tests.

## App preview:

- [Web](https://sg-web-rvt-ss-lang-diag.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/shared/src/search/query/diagnostics.test.ts [hunk 1]
```diff
@@ -607,7 +607,7 @@ describe('getDiagnostics()', () => {
 
     describe('structural search and type: filter', () => {
         test('detects type: filter in structural search', () => {
-            expect(parseAndDiagnose('type:symbol test', SearchPatternType.structural)).toMatchInlineSnapshot(`
+            expect(parseAndDiagnose('type:symbol test lang:go', SearchPatternType.structural)).toMatchInlineSnapshot(`
```
Comment (rvantonder) [inline client/shared/src/search/query/diagnostics.test.ts:610]: Just adding `lang:go` explicitly to these existing test so we can isolate the diagnostics being tested.

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.53 kb) | 0.00% (+0.53 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/rvt/ss-lang-diag/compare-report.html;tab=live_object) for a full comparison between the commits f26cdd0 and aba9c10 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Review by camdencheek — APPROVED** [review]
Nice, I've definitely been caught by this before

File: client/shared/src/search/query/diagnostics.ts [hunk 2]
```diff
@@ -246,13 +248,43 @@ const rules: PatternOf<Token[], PatternData>[] = [
...
+            some({ field: { value: 'patterntype' }, value: { value: 'structural' } }),
+            allOf(
+                (_tokens, context) => context.data.searchPatternType === SearchPatternType.structural,
+                not(some({ field: { value: 'patterntype' }, value: { value: not('structural') } }))
+            )
+        ),
+        oneOf({
+            $pattern: not(some({ field: { value: oneOf('language', 'lang', 'l') } })),
+            $data: addDiagnostic((_tokens, context) =>
+                context.data.patterns.length > 0
```
Comment (fkling) [inline client/shared/src/search/query/diagnostics.ts:264]: I _think_ you could just use `_tokens` to find the first pattern, i.e. you don't necessarily have to add `patterns` below, but I don't have strong feelings about this.

**Review by fkling — APPROVED** [review]

File: client/shared/src/search/query/diagnostics.ts (hunk 2 above)
Comment (rvantonder) [reply client/shared/src/search/query/diagnostics.ts:264]: Yeah, noticed that, I just decided to follow previous conventions

**Pull request merged**